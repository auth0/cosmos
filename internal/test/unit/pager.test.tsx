import { shallow } from 'enzyme'
import * as React from 'react'

import { Pager } from '@auth0/cosmos'

export type PagerFactoryParams = {
  page?: number
  perPage?: number
  items?: number
  showInfo?: boolean
  onPageChanged?: Function
}

const noop = () => {}

const paginationFactory = ({
  page,
  perPage,
  items,
  showInfo,
  onPageChanged = noop
}: PagerFactoryParams = {}) =>
  shallow(
    <Pager
      page={page}
      perPage={perPage}
      items={items}
      showInfo={showInfo}
      onPageChanged={onPageChanged}
    />
  )

describe('Pager', () => {
  describe('modes', () => {
    it('renders full mode', () => {
      const pagination = paginationFactory({ page: 4, perPage: 10, items: 10000 })

      expect(pagination).toMatchSnapshot()
    })

    it('renders page only mode', () => {
      const pagination = paginationFactory({ page: 4 })

      expect(pagination).toMatchSnapshot()
    })

    it('renders no info mode', () => {
      const pagination = paginationFactory({ page: 4, showInfo: false })

      expect(pagination).toMatchSnapshot()
    })

    it('renders previus page control disabled', () => {
      const pagination = paginationFactory({ page: 1, items: 10, perPage: 5 })

      expect(pagination).toMatchSnapshot()
    })

    it('renders next page control disabled', () => {
      const pagination = paginationFactory({ page: 2, items: 10, perPage: 5 })

      expect(pagination).toMatchSnapshot()
    })
  })
  describe('controls', () => {
    describe('previous page button', () => {
      it('calls onPageChanged when clicked', () => {
        const onPageChanged = jest.fn()
        const pagination = paginationFactory({ page: 4, perPage: 10, items: 10000, onPageChanged })
        const button = pagination.children().first()

        button.simulate('click')

        expect(onPageChanged).toHaveBeenCalledWith(3)
      })

      it('does not call onPageChanged when disabled and clicked', () => {
        const onPageChanged = jest.fn()
        const pagination = paginationFactory({ page: 1, perPage: 10, items: 10000, onPageChanged })
        const button = pagination.children().first()

        button.simulate('click')

        expect(onPageChanged).not.toHaveBeenCalled()
      })
    })

    describe('next page button', () => {
      it('calls onPageChanged when clicked', () => {
        const onPageChanged = jest.fn()
        const pagination = paginationFactory({ page: 4, perPage: 10, items: 10000, onPageChanged })
        const button = pagination.children().last()

        button.simulate('click')

        expect(onPageChanged).toHaveBeenCalledWith(5)
      })

      it('does not call onPageChanged when disabled and clicked', () => {
        const onPageChanged = jest.fn()
        const pagination = paginationFactory({ page: 4, perPage: 10, items: 40, onPageChanged })
        const button = pagination.children().last()

        button.simulate('click')

        expect(onPageChanged).not.toHaveBeenCalled()
      })
    })
  })
})
