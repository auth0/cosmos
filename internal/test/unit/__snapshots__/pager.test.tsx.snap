// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pager modes renders full mode 1`] = `
<styled.div
  data-cosmos-key="pager"
>
  <Button
    appearance="secondary"
    disabled={false}
    icon="chevron-left"
    iconAlign="left"
    loading={false}
    onClick={[Function]}
    size="compressed"
    success={false}
  >
    Newer
  </Button>
  <styled.div>
    Showing 31 - 40 of 10000
  </styled.div>
  <Button
    appearance="secondary"
    disabled={false}
    icon="chevron-right"
    iconAlign="right"
    loading={false}
    onClick={[Function]}
    size="compressed"
    success={false}
  >
    Older
  </Button>
</styled.div>
`;

exports[`Pager modes renders next page control disabled 1`] = `
<styled.div
  data-cosmos-key="pager"
>
  <Button
    appearance="secondary"
    disabled={false}
    icon="chevron-left"
    iconAlign="left"
    loading={false}
    onClick={[Function]}
    size="compressed"
    success={false}
  >
    Newer
  </Button>
  <styled.div>
    Showing 6 - 10 of 10
  </styled.div>
  <Button
    appearance="secondary"
    disabled={true}
    icon="chevron-right"
    iconAlign="right"
    loading={false}
    onClick={[Function]}
    size="compressed"
    success={false}
  >
    Older
  </Button>
</styled.div>
`;

exports[`Pager modes renders no info mode 1`] = `
<styled.div
  data-cosmos-key="pager"
>
  <Button
    appearance="secondary"
    disabled={false}
    icon="chevron-left"
    iconAlign="left"
    loading={false}
    onClick={[Function]}
    size="compressed"
    success={false}
  >
    Newer
  </Button>
  <styled.div />
  <Button
    appearance="secondary"
    disabled={false}
    icon="chevron-right"
    iconAlign="right"
    loading={false}
    onClick={[Function]}
    size="compressed"
    success={false}
  >
    Older
  </Button>
</styled.div>
`;

exports[`Pager modes renders page only mode 1`] = `
<styled.div
  data-cosmos-key="pager"
>
  <Button
    appearance="secondary"
    disabled={false}
    icon="chevron-left"
    iconAlign="left"
    loading={false}
    onClick={[Function]}
    size="compressed"
    success={false}
  >
    Newer
  </Button>
  <styled.div>
    Page 4
  </styled.div>
  <Button
    appearance="secondary"
    disabled={false}
    icon="chevron-right"
    iconAlign="right"
    loading={false}
    onClick={[Function]}
    size="compressed"
    success={false}
  >
    Older
  </Button>
</styled.div>
`;

exports[`Pager modes renders previus page control disabled 1`] = `
<styled.div
  data-cosmos-key="pager"
>
  <Button
    appearance="secondary"
    disabled={true}
    icon="chevron-left"
    iconAlign="left"
    loading={false}
    onClick={[Function]}
    size="compressed"
    success={false}
  >
    Newer
  </Button>
  <styled.div>
    Showing 1 - 5 of 10
  </styled.div>
  <Button
    appearance="secondary"
    disabled={false}
    icon="chevron-right"
    iconAlign="right"
    loading={false}
    onClick={[Function]}
    size="compressed"
    success={false}
  >
    Older
  </Button>
</styled.div>
`;
